### PR TITLE
Fix workflows when a child plan has no child tasks

### PR DESCRIPTION
**Pull request by bradyoo**
There were couple of bugs in the previous implementation:

1. If a plan has no children, the parent workflow currently doesn't get its status updated correctly because it currently relies on job workers reporting that the tasks are done/failed.

2. If a worker submits results for multiple tasks of the same job such that the job completes, workerHeartbeat on workflow crashes. This in particular causes https://github.com/Alluxio/alluxio/pull/10720 to break JobWorker heartbeats.

3. Workflows are not failed properly when job workers die.

This code fixes it by replacing the workerHeartbeat logic in workflow to being logic in the StatusChangeCallback to check for all changes in Status instead of some.

### CODE REVIEW

File: job/common/src/test/java/alluxio/job/DoNothingConfig.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+package alluxio.job;
+
+import alluxio.job.plan.PlanConfig;
+
+import com.google.common.base.MoreObjects;
+
+/**
+ * Config for a plan that does nothing.
+ */
+public class DoNothingConfig implements PlanConfig {
```
Comment (gpang) [inline job/common/src/test/java/alluxio/job/DoNothingConfig.java]: rename to `NoopPlanConfig` or `TestPlanConfig`

File: job/server/src/test/java/alluxio/job/plan/DoNothingPlanDefinition.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import alluxio.wire.WorkerInfo;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This definition of a plan that does nothing.
+ */
+public class DoNothingPlanDefinition
```
Comment (gpang) [inline job/server/src/test/java/alluxio/job/plan/DoNothingPlanDefinition.java]: `NoopPlanDefinition` of `TestPlanDefinition`

**Comment by bradyoo** [comment]
Addressed feedback

**Comment by bradyoo** [comment]
Abandoning this PR. Fixing it a different way.

**Comment by bradyoo** [comment]
Reopening now that implementation has changed. The implementation has been changed to handle all status changes.

**Comment by bradyoo** [comment]
Ping? @gpang or @calvinjia

**Comment by gpang** [comment]
@bradyoo could you briefly describe the fix/improvement in the PR description?

**Comment by bradyoo** [comment]
Description updated.

**Comment by bradyoo** [comment]
Ping? @gpang  or @calvinjia

File: job/common/src/test/java/alluxio/job/NoopPlanConfig.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import alluxio.job.plan.PlanConfig;
+
+import com.google.common.base.MoreObjects;
+
+/**
+ * Config for a plan that does nothing.
+ */
+public class NoopPlanConfig implements PlanConfig {
+
+  public static final String NAME = "DoNothing";
```
Comment (gpang) [inline job/common/src/test/java/alluxio/job/NoopPlanConfig.java]: should this be the same as the class name? `NoopPlanConfig`?

File: job/server/src/test/java/alluxio/job/plan/NoopPlanDefinition.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  public Set<Pair<WorkerInfo, SerializableVoid>> selectExecutors(NoopPlanConfig config,
+      List<WorkerInfo> jobWorkerInfoList, SelectExecutorsContext selectExecutorsContext)
+      throws Exception {
+    return new HashSet<>();
+  }
+
+  @Override
+  public SerializableVoid runTask(NoopPlanConfig config, SerializableVoid args,
+                                  RunTaskContext runTaskContext) throws Exception {
+    throw new IllegalStateException("Should never get here.");
```
Comment (gpang) [inline job/server/src/test/java/alluxio/job/plan/NoopPlanDefinition.java]: maybe say something like `NoopPlanDefinition should never run a task`

File: job/server/src/test/java/alluxio/job/plan/NoopPlanDefinition.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  @Override
+  public Class<NoopPlanConfig> getJobConfigClass() {
+    return NoopPlanConfig.class;
+  }
+
+  @Override
+  public Set<Pair<WorkerInfo, SerializableVoid>> selectExecutors(NoopPlanConfig config,
+      List<WorkerInfo> jobWorkerInfoList, SelectExecutorsContext selectExecutorsContext)
+      throws Exception {
+    return new HashSet<>();
```
Comment (gpang) [inline job/server/src/test/java/alluxio/job/plan/NoopPlanDefinition.java]: Is this allowed to return `Collections.emptySet()`?

File: job/server/src/main/java/alluxio/master/job/workflow/WorkflowTracker.java [hunk 6]
```diff
@@ -248,36 +248,24 @@ private synchronized void next(long jobId) throws ResourceExhaustedException {
       JobConfig childJobConfig = childJobConfigsIter.next();
       try {
         mJobMaster.run(childJobConfig, childJobId);
-      } catch (JobDoesNotExistException e) {
+      } catch (JobDoesNotExistException | ResourceExhaustedException e) {
         LOG.warn(e.getMessage());
```
Comment (gpang) [inline job/server/src/main/java/alluxio/master/job/workflow/WorkflowTracker.java]: would more info in this warning message be helpful?

File: job/server/src/main/java/alluxio/master/job/workflow/WorkflowTracker.java [hunk 7]
```diff
@@ -248,36 +248,24 @@ private synchronized void next(long jobId) throws ResourceExhaustedException {
...
-      }
-      Status status = jobInfo.getStatus();
-      if (status.equals(Status.COMPLETED)) {
-        done(planId);
-      } else if (status.equals(Status.CANCELED) || status.equals(Status.FAILED)) {
-        stop(planId, status, jobInfo.getErrorMessage());
-      }
+  public void onPlanStatusChange(PlanInfo planInfo) {
+    Status status = planInfo.getStatus();
+    if (status.equals(Status.COMPLETED)) {
```
Comment (gpang) [inline job/server/src/main/java/alluxio/master/job/workflow/WorkflowTracker.java]: should this be `switch` statement?

**Comment by bradyoo** [comment]
Reflected the feedback. (Though switch statement is significantly more wordy.) 

The JobDoesNotExistException and ResourceExhaustedException doesn't need much warning message because those 2 are error messages where the exception is very clear where and how it was thrown.

**Comment by gpang** [comment]
alluxio-bot, merge this please

**Pull request merged**